state/oracledatabase/oracledatabaseaccess.go

```diff
@@ -0,0 +1,377 @@
+/*
+Copyright 2021 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package oracledatabase
+
+import (
+	"database/sql"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"strconv"
+
+	"github.com/google/uuid"
+
+	"github.com/dapr/components-contrib/state"
+	"github.com/dapr/components-contrib/state/utils"
+	"github.com/dapr/kit/logger"
+
+	// Blank import for the underlying Oracle Database driver.
+	_ "github.com/sijms/go-ora/v2"
+)
+
+const (
+	connectionStringKey        = "connectionString"
+	oracleWalletLocationKey    = "oracleWalletLocation"
+	metadataTTLKey             = "ttlInSeconds"
+	errMissingConnectionString = "missing connection string"
+	tableName                  = "state"
+)
+
+// oracleDatabaseAccess implements dbaccess.
+type oracleDatabaseAccess struct {
+	logger           logger.Logger
+	metadata         state.Metadata
+	db               *sql.DB
+	connectionString string
+	tx               *sql.Tx
+}
+
+// newOracleDatabaseAccess creates a new instance of oracleDatabaseAccess.
+func newOracleDatabaseAccess(logger logger.Logger) *oracleDatabaseAccess {
+	logger.Debug("Instantiating new Oracle Database state store")
+
+	return &oracleDatabaseAccess{
+		logger: logger,
+	}
+}
+
+// Init sets up OracleDatabase connection and ensures that the state table exists.
+func (o *oracleDatabaseAccess) Init(metadata state.Metadata) error {
+	o.logger.Debug("Initializing OracleDatabase state store")
+	o.metadata = metadata
+	if val, ok := metadata.Properties[connectionStringKey]; ok && val != "" {
+		o.connectionString = val
+	} else {
+		o.logger.Error("Missing Oracle Database connection string")
+
+		return fmt.Errorf(errMissingConnectionString)
+	}
+	if val, ok := o.metadata.Properties[oracleWalletLocationKey]; ok && val != "" {
+		o.connectionString += "?TRACE FILE=trace.log&SSL=enable&SSL Verify=false&WALLET=" + url.QueryEscape(val)
+	}
+	db, err := sql.Open("oracle", o.connectionString)
+	if err != nil {
+		o.logger.Error(err)
+
+		return err
+	}
+
+	o.db = db
+
+	if pingErr := db.Ping(); pingErr != nil {
```

Comment:
Now `Ping()` always return nil, how about comment this logic and add TODO?

Reply:
Well spotted, an omission on my part. I have now implemented the Ping method - also in the mock object - and added a unit test.